### ISSUE TITLE
RFE: explicitly output build directory path(s) in the build log

### ISSUE THREAD

**Issue by pmatilai** [rpm-software-management/rpm#3147]
With the introduction of the new intermediate builddir in 4.20, we're running into multiple cases where people/scripts/external tooling expects to manipulate stuff in the rpm build directory after performing a partial build on it, eg 'rpmbuild -bp' to have rpm prepare the sources.

Rpm never gave any promises about the directory outside itself, but when a thing stays the same for 20+ years people start assuming it, fact of life. Those use-cases will not go away so we should instead help them on our own terms: we do not want to make promises about build paths outside a build context, but it's absolutely no problem to tell others when we're actually doing it.

There are several closely related directories that we might want to explicitly state in the build log, such as %{builddir}, the *actual* build directory for %setup'ed packages which is %{builddir}/%{?buildsubdir} (we don't even have a good name for it), and buildroot.

**Comment by pmatilai** [comment]
Another related idea is that we could maybe output the expanded `%___build_pre_env` contents into file with a known name such as rpmbuild.env, and source that from the scriptlets instead of re-expanding it every time. The build directories and already in there in a shell consumable form, and would be useful for somebody trying to run things the rpm-way from the "outside".

**Comment by dmnks** [comment]
Yet another idea: Have an rpmbuild subcommand that runs an interactive shell (or a script if supplied instead) *in* the same environment. Basically, what you propose but flipped inside out :smile: 

E.g. `rpmbuild --shell sh -c 'echo $RPM_BUILD_DIR'`

**Comment by lukaszachy** [comment]
`rpmbuild.env`-like approach seems great for our https://github.com/teemtee/tmt/issues/2987 uses case as we just want to know where to look for files (content of $RPM_BUILD_DIR).

**Comment by pmatilai** [comment]
> Yet another idea: Have an rpmbuild subcommand that runs an interactive shell (or a command if supplied instead) _in_ the same environment. Basically, what you propose but flipped inside out 😄
> 
> E.g. `rpmbuild --shell sh -c 'echo $RPM_BUILD_DIR'`

Heh, I had this same exact idea yesterday evening after closing the laptop for the day.
We indeed already have similar things around, rpmspec and rpmlua have --shell, and like you said, our test-suite has it too.

Because the needs of folks running stuff this way vary a lot, I think --shell would need to be a "subcommand" that runs before exiting rpmbuild. Eg,

'rpmbuild -bp --shell ...' would run the shell bit after running %prep, but before exiting, 'rpmbuild -bc --shell' after compile phase, etc. It probably should be disabled during actual package builds.

As for printing out relevant parts, instead of adding individual prints with names in there, I think we should just dump the (rpm specific) environment out into the build log. It'll help diagnosis in several other situations too, right now you need to catch an in-flight scriptlet execution to see what exactly is in the temporary scriptlet that is actually executing. And, if we put the environment into a file, which has the benefit of being sourceable for other purposes, printing it out is just a matter of simple "cat".

This is starting to look like a plan with several benefits and no downsides.

**Comment by pmatilai** [comment]
In the meanwhile, here's a trick how you can fish out this information from existing rpmbuild, old and new, without even unpacking sources:

> rpmbuild -bc --noprep --define '__spec_build_pre echo builddir=%{_builddir}; exit 0' rpm.spec |awk /^builddir/

This gives you the toplevel directory where setup will unpack the sources, ie what the existing test systems are build around, ie: 
- in rpm < 4.19 it's ~/rpmbuild/BUILD/
- in rpm >= 4.20 it's *currently* something like (please do not hardcode this!) ~/rpmbuild/BUILD/<name>-<version>-build

**Comment by dmnks** [comment]
> Heh, I had this same exact idea yesterday evening after closing the laptop for the day.

Oh :smile: Good, it's not a completely crazy idea then.

I didn't think it through too much, the use case just felt familiar enough, mostly due to the recent test-suite rework but of course it's a common theme with other similar tooling, as you said.

> We indeed already have similar things around, rpmspec and rpmlua have --shell, and like you said, our test-suite has it too.

Yup, somehow forgot about those utilities we ourselves ship :smile: And then there's `mock --shell`, `mkosi shell`, etc.

> Because the needs of folks running stuff this way vary a lot, I think --shell would need to be a "subcommand" that runs before exiting rpmbuild.

Indeed, what we want here basically is to allow the user to swap out the respective build script with their own shell (script) that executes in the exact same environment, thus giving them access to all of its internals for their own purposes, including just `echo`ing the env vars.

> Eg, 'rpmbuild -bp --shell ...' would run the shell bit after running %prep, but before exiting, 'rpmbuild -bc --shell' after compile phase, etc. It probably should be disabled during actual package builds.

Oh, good idea, yup. Reusing the existing stage options sounds natural and wouldn't require additional options.

> As for printing out relevant parts, instead of adding individual prints with names in there, I think we should just dump the (rpm specific) environment out into the build log. It'll help diagnosis in several other situations too, right now you need to catch an in-flight scriptlet execution to see what exactly is in the temporary scriptlet that is actually executing. And, if we put the environment into a file, which has the benefit of being sourceable for other purposes, printing it out is just a matter of simple "cat".

Right, inspecting a build script would be possible with the `--shell` option already, but having a simple way to get the actual script that gets executed by rpmbuild without having to hunt for the temp file (as you said) would indeed be nice (and something I have wished for on many occasions myself :smile:).

> This is starting to look like a plan with several benefits and no downsides.

Indeed. These are basically two separate features at this point, although somewhat connected.

**Comment by dmnks** [comment]
> Indeed, what we want here basically is to allow the user to swap out the respective build script as defined in the spec with their own shell (script) that executes in the exact same environment, thus giving them access to all of its internals for their own purposes, including just `echo`ing the env vars.

Correction: We don't want them to *replace* the script of course, just *extend* it (as you noted).

**Comment by pmatilai** [comment]
Yeah and actually with --shell, what's executed can be an entirely independent from what's in the spec, its just something that is executed in the environment.

**Comment by dmnks** [comment]
I guess that depends on whether we want `--shell` to execute in place of the respective scriptlet (its definition in the spec) or after its execution.

**Comment by pmatilai** [comment]
I tend to think the latter, but it needs more consideration on the actual cases people would/could use it for. And, I agree it's really a separate topic, should probably open another ticket for it.

**Comment by omajid** [comment]
The workaround described above (`rpmbuild -bc --short-circuit ...`) can't be easily combined with tools like fedpkg. I filed https://pagure.io/fedpkg/issue/551
